### ISSUE TITLE
`scatter_sum` leading to suboptimal performance

### ISSUE THREAD

**Issue by MartinXPN** [rusty1s/pytorch_geometric#2306]
## 🐛 Bug

I've been trying to profile the performance of my model to see where it spends the most time and what could be optimized. I've used `pyinstrument` to get the statistics. The performance was measured on a CPU and on large graphs where the number of nodes and edges are greater than 1 million. The model consists of several GCNConv blocks stacked together along with PReLU, Dropouts, and several Linear layers. The heaviest part from what I saw was the `torch_scatter/scatter` called in `message_passing`. 
<p align="center">
  <img width="700" alt="Screen Shot 2021-03-25 at 10 16 02 AM" src="https://user-images.githubusercontent.com/16335716/112428663-f315e600-8d54-11eb-934e-a72b25d5000c.png">
</p>

Then I tried to run the prediction script without JIT so that I can see more verbose logging of the performance (several functions are omitted in the report when using `torch.jit`). The experiment showed that `message_passing.aggregate` &rarr; `scatter. scatter` &rarr; `scatter. scatter_sum` &rarr; `Tensor.scatter_add_` was the most demanding part.
<p align="center">
  <img width="700" alt="Screen Shot 2021-03-25 at 10 16 43 AM" src="https://user-images.githubusercontent.com/16335716/112429040-8fd88380-8d55-11eb-833e-9c4b7d9648cb.png">
</p>

After playing around with the code I realized that the index is broadcasted with
```python
index = broadcast(index, src, dim)
...
...
# after which performing 
out = torch.zeros(size, dtype=src.dtype, device=src.device)
return out.scatter_add_(dim, index, src)
```
consumes both a lot of memory and does not parallelize. `scatter_add_` is run on only a single core. 

## To Reproduce

1. Create a model consisting of several stacked GCNConvs
1. Take a big graph `|V| > 1M`, `|W| > 1M`
1. Run the script with `pyinstrument`

## Expected behavior
The expected behavior would be to run the `scatter_add_` on several cores and parallelize it fully without consuming too much memory because of the broadcast of the `index`.

## Environment

* OS: MacOS/Linux
* Python version: 3.8.6
* PyTorch version: 1.7.0
* CUDA/cuDNN version: -
* GCC version: 4.2.1
* Any other relevant information: CPU version of pytorch geometric

**Comment by MartinXPN** [comment]
I've tried to optimize the `scatter_sum` function by replacing it with something like this: (where I loop through all the items in `last_dim_size` instead of broadcasting the index and parallelize it with `torch.jit.fork`.
```python
@torch.jit.script
def foo(size: List[int], src: torch.Tensor, dim: int, index: torch.Tensor):
    return torch.zeros(size, dtype=src.dtype, device=src.device).scatter_add_(dim, index, src)

@torch.jit.script
def scatter_sum_optimized(src: torch.Tensor, index: torch.Tensor, dim: int = -1,
                out: Optional[torch.Tensor] = None,
                dim_size: Optional[int] = None) -> torch.Tensor:
    if out is not None and dim != -2:
        raise NotImplementedError

    # index = broadcast(index, src, dim)

    size = src.size()
    size = list(size)
    if dim_size is not None:
        size[dim] = dim_size
    elif index.numel() == 0:
        size[dim] = 0
    else:
        size[dim] = int(index.max()) + 1

    last_dim_size = size[-1]
    size[-1] = 1

    futures = [
        torch.jit.fork(foo, size=size, src=src[:, i: i + 1], dim=dim, index=index.unsqueeze(dim=-1))
        for i in range(last_dim_size)
    ]

    return torch.cat([
        torch.jit.wait(future) for future in futures
    ], dim=-1)
```

This improved the performance several times!

I'd love to create a PR with a fix.
But I guess there could be a better solution
* A better place to fix it in the code maybe instead of rewriting `scatter_sum`
* A more generic way to fix this as now I've only handled the 2D version where I know for sure the `index` was broadcasted
* I'm not sure if this breaks the training as I've only tried running it in inference mode

**Comment by rusty1s** [comment]
That's an interesting solution. It is true that `scatter_*` on CPU currently only runs on a single core. We have plans to integrate that functionality into `torch-scatter` directly. If you are interested, we can work on this together.

In particular, we want to replace the [current implementation](https://github.com/rusty1s/pytorch_scatter/blob/master/csrc/cpu/scatter_cpu.cpp#L65-L79) with a parallel implementation that makes use of the `at::parallel_for` interface of `libtorch`, such as [we do here](https://github.com/rusty1s/pytorch_cluster/blob/master/csrc/cpu/rw_cpu.cpp#L14-L15).

**Comment by MartinXPN** [comment]
Yeah that sounds great! I can try to help with the `scatter_*` implementation and create a PR.
